Extend/Payment_Gateway.php

```diff
@@ -0,0 +1,266 @@
+        $oTransaction = $oPaymentMethod->getTransaction();
+        $oTransaction->setRedirect($oRedirect);
+
+        $config = \OxidEsales\Eshop\Core\Registry::getConfig();
+        $currency = $config->getActShopCurrencyObject();
+        $oTransaction->setAmount(new Amount($dAmount, $currency->name));
+
+        $basket = $oSession->getBasket();
+        $user = $basket->getBasketUser();
+
+        $oTransaction->setOrderDetail(sprintf(
+            '%s %s %s',
+            $oOrder->oxorder__oxbillemail->value,
+            $user->oxuser__oxfname->value,
+            $user->oxuser__oxlname->value
+        ));
+
+        $payment = oxNew(\OxidEsales\Eshop\Application\Model\Payment::class);
+        $payment->load(self::NAME);
+
+        if ($payment->oxpayments__wdoxidee_additional_info->value) {
+            $oViewConf = oxNew('oxViewConfig');
+            $ip = $oViewConf->getRemoteAddress();
+            $oTransaction->setIpAddress($ip);
+            $oTransaction->setConsumerId($user->oxuser__oxid->value);
+            $oTransaction->setOrderNumber($oOrder->oxorder__oxid->value);
+        }
+
+        if ($payment->oxpayments__wdoxidee_descriptor->value) {
+            $descriptor = self::getDescriptor($oOrder->oxorder__oxid->value);
+            $oTransaction->setDescriptor($descriptor);
+        }
+
+        $address = new \Wirecard\PaymentSdk\Entity\Address(
+            self::getCountryCode($user->oxuser__oxcountryid),
+            $user->oxuser__oxcity->value,
+            $user->oxuser__oxstreet->value
+        );
+        $address->setPostalCode($user->oxuser__oxzip->value);
+
+        $accountHolder = new \Wirecard\PaymentSdk\Entity\AccountHolder();
+        $accountHolder->setAddress($address);
+        $accountHolder->setFirstName($user->oxuser__oxfname->value);
+        $accountHolder->setLastName($user->oxuser__oxlname->value);
+        $accountHolder->setPhone($user->oxuser__oxfon->value);
+        $accountHolder->setEmail($oOrder->oxorder__oxbillemail->value);
+
+        $oTransaction->setAccountHolder($accountHolder);
+
+        if ($oOrder->oxorder__oxdelfname->value) { //shipping address exists
+            $addressShipping = new \Wirecard\PaymentSdk\Entity\Address(
+                self::getCountryCode($oOrder->oxorder__oxdelcountryid),
+                $oOrder->oxorder__oxdelcity->value,
+                $oOrder->oxorder__oxdelstreet->value
+            );
+            $addressShipping->setPostalCode($oOrder->oxorder__oxdelzip->value);
+            $accountHolderShipping = new \Wirecard\PaymentSdk\Entity\AccountHolder();
+            $accountHolderShipping->setAddress($addressShipping);
+            $accountHolderShipping->setFirstName($oOrder->oxorder__oxdelfname->value);
+            $accountHolderShipping->setLastName($oOrder->oxorder__oxdellname->value);
+            $accountHolderShipping->setPhone($oOrder->oxorder__oxdelfon->value);
+            $oTransaction->setShipping($accountHolderShipping);
+        } else {
```

Comment:
please during the refactoring of this method, build the methods so that you will not need this else.

Reply:
this cannot be split, the values come from different tables and columns.

Reply:
the method is to long to much logic is in it and with the splinting of the method i suggest you that you write the code so that you don't need an else.

Reply:
of course, we split the `makeTransaction` but we cannot get rid of the else for the address.

Reply:
This is just a 5min example you can refactor it even better. I think it is possible.
```php

public function makeTransaction() {
    //for paypal you set both shipping and billing
    $this->setAddresses($transaction);

}

public function setAddresses($transaction, $addShipping = true) {
    $accountHolder = $this->buildAccountHolder();
    $transaction->setAccounHolder(
        $accountHolder
    );
    if ($this->isSame()) {
        $transaction->setShipping(
            $accountHolder
        );
        return $transaction;
    }
    $shipping = $this->buildShipping();
    $transaction->setShipping($shipping);

    return $transaction;
}
```